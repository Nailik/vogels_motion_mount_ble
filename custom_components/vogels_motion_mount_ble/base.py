"""Base entity which all other entity platform classes can inherit.

As all entity types have a common set of properties, you can
create a base entity like this and inherit it in all your entity platforms.

This just makes your code more efficient and is totally optional.

See each entity platform (ie sensor.py, switch.py) for how this is inheritted
and what additional properties and methods you need to add for each entity type.

"""

import logging
from typing import Any

from homeassistant.core import callback
from homeassistant.helpers.device_registry import DeviceInfo
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from .const import DOMAIN
from .coordinator import ExampleCoordinator

_LOGGER = logging.getLogger(__name__)


class ExampleBaseEntity(CoordinatorEntity):
    """Base Entity Class.

    This inherits a CoordinatorEntity class to register your entites to be updated
    by your DataUpdateCoordinator when async_update_data is called, either on the scheduled
    interval or by forcing an update.
    """

    coordinator: ExampleCoordinator

    # ----------------------------------------------------------------------------
    # Using attr_has_entity_name = True causes HA to name your entities with the
    # device name and entity name.  Ie if your name property of your entity is
    # Voltage and this entity belongs to a device, Lounge Socket, this will name
    # your entity to be sensor.lounge_socket_voltage
    #
    # It is highly recommended (by me) to use this to give a good name structure
    # to your entities.  However, totally optional.
    # ----------------------------------------------------------------------------
    _attr_has_entity_name = True

    def __init__(self, coordinator: ExampleCoordinator) -> None:
        """Initialise entity."""
        super().__init__(coordinator)

    @property
    def device_info(self) -> DeviceInfo:
        """Return device information."""

        # ----------------------------------------------------------------------------
        # Identifiers are what group entities into the same device.
        # If your device is created elsewhere, you can just specify the indentifiers
        # parameter to link an entity to a device.
        # If your device connects via another device, add via_device parameter with
        # the indentifiers of that device.
        #
        # Device identifiers should be unique, so use your integration name (DOMAIN)
        # and a device uuid, mac address or some other unique attribute.
        # ----------------------------------------------------------------------------
        return DeviceInfo(
            name=self.coordinator.name,
            manufacturer="Vogelsr",
            model="Motion Mount",
            identifiers={(DOMAIN, self.coordinator.mac)},
        )

    @property
    def name(self) -> str:
        """Return the name of the sensor."""
        return self.coordinator.name

    @property
    def unique_id(self) -> str:
        """Return unique id."""

        # ----------------------------------------------------------------------------
        # All entities must have a unique id across your whole Home Assistant server -
        # and that also goes for anyone using your integration who may have many other
        # integrations loaded.
        #
        # Think carefully what you want this to be as changing it later will cause HA
        # to create new entities.
        #
        # It is recommended to have your integration name (DOMAIN), some unique id
        # from your device such as a UUID, MAC address etc (not IP address) and then
        # something unique to your entity (like name - as this would be unique on a
        # device)
        #
        # If in your situation you have some hub that connects to devices which then
        # you want to create multiple sensors for each device, you would do something
        # like.
        #
        # f"{DOMAIN}-{HUB_MAC_ADDRESS}-{DEVICE_UID}-{ENTITY_NAME}""
        #
        # This is even more important if your integration supports multiple instances.
        # ----------------------------------------------------------------------------
        return f"{DOMAIN}-{self.coordinator.mac}"
