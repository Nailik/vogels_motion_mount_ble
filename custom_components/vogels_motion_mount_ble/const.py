"""Constants for the Vogels Motion Mount BLE integration."""

DOMAIN = "vogels_motion_mount_ble"

SERVICE_UUID = "3e6fe65d-ed78-11e4-895e-00026fd5c52c"

CHAR_NAME_UUID = "c005fa37-0651-4800-b000-000000000000"
CHAR_SIZE_UUID = "c005fa2b-0651-4800-b000-000000000000"
CHAR_DISTANCE_UUID = "c005fa00-0651-4800-b000-000000000000"
CHAR_ROTATION_UUID = "c005fa01-0651-4800-b000-000000000000"
CHAR_AUTOMATIC_UUID = "c005fa02-0651-4800-b000-000000000000"
CHAR_FREEZE_UUID = "c005fa14-0651-4800-b000-000000000000"
CHAR_PRESET_UUIDS = [
    "c005fa0a-0651-4800-b000-000000000000",
    "c005fa0b-0651-4800-b000-000000000000",
    "c005fa0c-0651-4800-b000-000000000000",
    "c005fa0d-0651-4800-b000-000000000000",
    "c005fa0e-0651-4800-b000-000000000000",
    "c005fa0f-0651-4800-b000-000000000000",
    "c005fa10-0651-4800-b000-000000000000",
]
CHAR_PRESET_UUID = "c005fa2a-0651-4800-b000-000000000000"
