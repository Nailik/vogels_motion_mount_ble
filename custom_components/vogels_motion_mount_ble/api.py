"""API Placeholder.

You should create your api seperately and have it hosted on PYPI.  This is included here for the sole purpose
of making this example code executable.
"""

import asyncio
from dataclasses import dataclass, replace
import logging
from .const import CHAR_DISTANCE_UUID, CHAR_ROTATION_UUID, CHAR_PRESET_UUID
from bleak import BleakClient, BleakScanner

_LOGGER = logging.getLogger(__name__)

@dataclass
class VogelsMotionMountData:
    """Class to hold api data."""

    connected: bool = False
    distance: int | None = None
    rotation: int | None = None

class API:
    """Class for example API."""

    def __init__(self, mac: str, pin: str | None, coordinator) -> None:
        """Initialise."""
        self._mac = mac
        self._pin = pin
        self._coordinator = coordinator
        self._data = VogelsMotionMountData(connected=False)
        self._client: BleakClient | None = None
        self._disconnected_event = asyncio.Event()

    async def testConnect(self):
        """Test connection to the BLE device once using BleakClient."""
        try:
            _LOGGER.debug("start test connection")
            async with BleakClient(self._mac, timeout=120) as client:
                _LOGGER.debug("Connected to device: %s", client.is_connected)
        except Exception as err:
            _LOGGER.error("Failed to connect to device: %s", err)
            raise APIConnectionError("Error connecting to api.") from err


    async def perform_action(self):
        _LOGGER.exception("write_gatt_char")
        result = await self._client.write_gatt_char(CHAR_PRESET_UUID, bytes([1]), response=True)
        _LOGGER.exception("write_gatt_char result: %s", result)

    def update(self, **kwargs):
        """Update one or more fields, retaining others from existing data. Then notify the coordinator."""
        self._data = replace(self._data, **kwargs)
        self._coordinator.async_set_updated_data(self._data)

    def _handle_disconnect(self, _):
        _LOGGER.debug("handle Device disconnected!")
        self._disconnected_event.set()

    def _handle_distance_change(self, _, data):
        _LOGGER.debug("distance change %s", data)
        self.update(distance=int.from_bytes(data, "little"))

    def _handle_rotation_change(self, _, data):
        _LOGGER.debug("rotation change %s", data)
        self.update(rotation=int.from_bytes(data, "little"))

    # todo make it optional if the connection should be maintained or connect on command (when required to send command) or poll time
    async def maintainConnection(self):
        """Maintain connection to device."""
        while True:
            _LOGGER.debug("scanning for device %s", self._mac)
            device = await BleakScanner.find_device_by_address(self._mac)

            if device is None:
                _LOGGER.debug("no device %s found, wait then scan again", self._mac)
                await asyncio.sleep(5)
                continue

            try:
                _LOGGER.debug("connecting to device")
                async with BleakClient(device, disconnected_callback=self._handle_disconnect, timeout=120) as client:
                    self.connected = client.is_connected
                    self._client = client
                    _LOGGER.debug("maintain connected to device %s", self._client.is_connected)
                    self.update(connected = self._client.is_connected)

                    self._handle_distance_change(None, await client.read_gatt_char(CHAR_DISTANCE_UUID))
                    self._handle_rotation_change(None, await client.read_gatt_char(CHAR_ROTATION_UUID))

                    await client.start_notify(CHAR_DISTANCE_UUID, self._handle_distance_change)
                    await client.start_notify(CHAR_ROTATION_UUID, self._handle_rotation_change)

                    await self._disconnected_event.wait()
                    # reset event
                    self._disconnected_event.clear()
                    _LOGGER.debug("device disconnected %s", self.connected)
                    self.update(connected = client.is_connected)
                    await asyncio.sleep(1)
            except Exception:
                # catch bleak.exc.BleakError: No backend with an available connection slot that can reach address D9:13:5D:AB:3B:37 was found
                _LOGGER.exception("Exception while connecting/connected")
                await asyncio.sleep(5)


class APIConnectionError(Exception):
    """Exception class for connection error."""
