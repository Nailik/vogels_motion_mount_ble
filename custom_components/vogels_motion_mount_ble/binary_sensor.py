"""Interfaces with the Integration 101 Template api sensors."""

import logging

from homeassistant.components.binary_sensor import (
    BinarySensorDeviceClass,
    BinarySensorEntity,
)
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from . import MyConfigEntry
from .base import ExampleBaseEntity
from .const import DOMAIN
from homeassistant.helpers import config_validation as cv
import voluptuous as vol
from homeassistant.helpers.typing import ConfigType
from homeassistant.core import HomeAssistant, ServiceCall, callback

_LOGGER = logging.getLogger(__name__)

async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: MyConfigEntry,
    async_add_entities: AddEntitiesCallback,
):
    """Set up the Binary Sensors."""
    coordinator = hass.data[DOMAIN][config_entry.entry_id]
    async_add_entities([ExampleBinarySensor(coordinator)])


class ExampleBinarySensor(ExampleBaseEntity, BinarySensorEntity):
    """Implementation of a sensor."""

    _attr_device_class = BinarySensorDeviceClass.CONNECTIVITY
    _attr_has_entity_name = True
    _attr_translation_key = "connection"
    _attr_should_poll = True

    @callback
    def _handle_coordinator_update(self) -> None:
        """Update sensor with latest data from coordinator."""
        _LOGGER.exception("_handle_coordinator_update binary sensor")
        # This method is called by your DataUpdateCoordinator when a successful update runs.
        self.async_write_ha_state()

    @property
    def name(self) -> str:
        """Return the name of the sensor."""
        return "connection"

    @property
    def unique_id(self) -> str:
        """Return unique id."""
        # All entities must have a unique id.  Think carefully what you want this to be as
        # changing it later will cause HA to create new entities.
        return f"{DOMAIN}-{self.coordinator.mac}-connected"

    @property
    def is_on(self) -> bool:
        """Return if the binary sensor is on."""
        _LOGGER.exception("is_on binary sensor")
        if not self.coordinator.data:
            _LOGGER.error("No data available in coordinator")
            return None
        # This needs to enumerate to true or false
        return self.coordinator.data.connected

