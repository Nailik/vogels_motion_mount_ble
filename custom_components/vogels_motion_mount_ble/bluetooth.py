from bleak import BleakClient, BleakScanner
import logging
import asyncio
from .const import CHAR_DISTANCE_UUID, CHAR_ROTATION_UUID, CHAR_PRESET_UUIDS
from homeassistant.core import HomeAssistant
from homeassistant.config_entries import ConfigEntry
from homeassistant.helpers.entity_platform import AddEntitiesCallback

logger = logging.getLogger(__name__)


async def async_setup_entry(
    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
):
    # Create an instance of your sensor entity
    handler = BluetoothDeviceHandler(entry.data["mac_address"], hass)
    # Add entity to Home Assistant
    async_add_entities(handler)


class BluetoothDeviceHandler:
    def __init__(self, address, hass):
        self._attr_name = "Bluetooth Connection Status"
        self._attr_unique_id = f"bluetooth_{address.replace(':', '').lower()}"
        self._address = address
        self._connected = False
        self._hass = hass
        self._client = None
        self._task = None
        self._disconnect_event = None

    def register(self):
        logger.debug("register")
        self._task = self._hass.loop.create_task(self.monitor())

    async def monitor(self):
        while True:
            logger.debug("scanning for device")
            device = await BleakScanner.find_device_by_address(self._address)

            if device is None:
                logger.debug("no device found, wait then scan again")
                await asyncio.sleep(5)
                continue

            disconnect_event = asyncio.Event()

            try:
                logger.debug("connecting to device")
                async with BleakClient(
                    device, disconnected_callback=disconnect_event.set()
                ) as client:
                    logger.debug("connected to device")
                    self._client = client
                    self._connected = True
                    await disconnect_event.wait()
                    logger.debug("device disconnected")
                    self._connected = False
            except Exception:
                logger.exception("exception while connecting/connected")
                await asyncio.sleep(5)

    @property
    def device_info(self):
        return {
            "identifiers": {("vogels_motion_mount", self._mac_address)},
            "manufacturer": "Custom Bluetooth Inc.",
            "model": "BT Device",
            "connections": {("mac", self._mac_address)},
        }

    @property
    def native_value(self):
        return self._connected

    @property
    def extra_state_attributes(self):
        return {
            "mac_address": self._address,
            "connected": self._connected,
        }

    async def is_connected(self):
        return self.client.is_connected

    async def get_distance(self):
        data = await self.client.read_gatt_char(CHAR_DISTANCE_UUID)
        return int.from_bytes(data[0:2], "little")

    async def set_distance(self, value: int):
        await self.client.write_gatt_char(
            CHAR_DISTANCE_UUID, value.to_bytes(2, "little")
        )

    async def get_rotation(self):
        data = await self.client.read_gatt_char(CHAR_ROTATION_UUID)
        return int.from_bytes(data[0:2], "little", signed=True)

    async def set_rotation(self, value: int):
        await self.client.write_gatt_char(
            CHAR_ROTATION_UUID, value.to_bytes(2, "little", signed=True)
        )

    async def get_presets(self):
        presets = []
        for uuid in CHAR_PRESET_UUIDS:
            raw = await self.client.read_gatt_char(uuid)
            if len(raw) < 4:
                continue
            distance = int.from_bytes(raw[0:2], "little")
            rotation = int.from_bytes(raw[2:4], "little", signed=True)
            name = raw[4:].decode("utf-8", errors="ignore").strip("\x00").strip()
            presets.append(
                {"uuid": uuid, "distance": distance, "rotation": rotation, "name": name}
            )
        return presets

    async def async_will_remove_from_hass(self):
        if self._task:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass
