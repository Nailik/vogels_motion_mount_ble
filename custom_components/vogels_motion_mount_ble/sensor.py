"""Interfaces with the Integration 101 Template api sensors."""

import logging

from homeassistant.components.sensor import (
    SensorDeviceClass,
    SensorEntity,
    SensorStateClass,
)
from homeassistant.const import (
    UnitOfElectricCurrent,
    UnitOfElectricPotential,
    UnitOfEnergy,
    UnitOfTemperature,
)
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from . import MyConfigEntry
from .base import ExampleBaseEntity
from .const import DOMAIN
from .coordinator import ExampleCoordinator

_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: MyConfigEntry,
    async_add_entities: AddEntitiesCallback,
):
    """Set up the Sensors."""
    # This gets the data update coordinator from the config entry runtime data as specified in your __init__.py
    coordinator: ExampleCoordinator = config_entry.runtime_data.coordinator

    # Enumerate all the sensors in your data value from your DataUpdateCoordinator and add an instance of your sensor class
    # to a list for each one.
    # This maybe different in your specific case, depending on how your data is structured
    sensors = [DistanceSensor(coordinator),RotationSensor(coordinator)]

    # Create the sensors.
    async_add_entities(sensors)


class DistanceSensor(ExampleBaseEntity, SensorEntity):
    """Implementation of a sensor."""

    _attr_device_class = SensorDeviceClass.DISTANCE

    @property
    def name(self) -> str:
        """Return the name of the sensor."""
        return "distance"

    @property
    def unique_id(self) -> str:
        """Return unique id."""
        # All entities must have a unique id.  Think carefully what you want this to be as
        # changing it later will cause HA to create new entities.
        return f"{DOMAIN}-{self.coordinator.mac}-distance"

    @property
    def native_value(self) -> int:
        """Return the state of the entity."""
        _LOGGER.exception("native_value DistanceSensor")
        if not self.coordinator.data:
            _LOGGER.error("No data available in coordinator")
            return None
        # This needs to enumerate to true or false
        return self.coordinator.data.distance

class RotationSensor(ExampleBaseEntity, SensorEntity):
    """Implementation of a sensor."""

    _attr_device_class = None
    _attr_native_unit_of_measurement = "Â°"

    @property
    def name(self) -> str:
        """Return the name of the sensor."""
        return "rotation"

    @property
    def unique_id(self) -> str:
        """Return unique id."""
        # All entities must have a unique id.  Think carefully what you want this to be as
        # changing it later will cause HA to create new entities.
        return f"{DOMAIN}-{self.coordinator.mac}-rotation"

    @property
    def native_value(self) -> int:
        """Return the state of the entity."""
        _LOGGER.exception("native_value RotationSensor")
        if not self.coordinator.data:
            _LOGGER.error("No data available in coordinator")
            return None
        # This needs to enumerate to true or false
        return self.coordinator.data.rotation
