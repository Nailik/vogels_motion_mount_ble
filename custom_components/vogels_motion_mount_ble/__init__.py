"""Integration for a Vogels Motion Mount via BLE."""

from __future__ import annotations

import logging
from homeassistant.const import __version__ as ha_version
from packaging import version
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import Platform
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import (
    ConfigEntryAuthFailed,
    ConfigEntryError,
    ConfigEntryNotReady,
)

from .api import APIAuthenticationError, APIConnectionDeviceNotFoundError
from .coordinator import VogelsMotionMountBleCoordinator
from .services import async_setup_services

_LOGGER = logging.getLogger(__name__)

PLATFORMS: list[Platform] = [
    Platform.BINARY_SENSOR,
    Platform.BUTTON,
    Platform.NUMBER,
    Platform.SELECT,
    Platform.SENSOR,
    Platform.SWITCH,
    Platform.TEXT,
]

type VogelsMotionMountBleConfigEntry = ConfigEntry[VogelsMotionMountBleCoordinator]

# for this version read_only was added to config selectors
MIN_HA_VERSION = "2025.6.0"

if version.parse(ha_version) < version.parse(MIN_HA_VERSION):
    raise RuntimeError(
        f"Vogels Motion Mount BLE requires Home Assistant {MIN_HA_VERSION}+"
    )


async def async_setup(
    hass: HomeAssistant, entry: VogelsMotionMountBleConfigEntry
) -> bool:
    """Set up Vogels Motion Mount integration services."""
    _LOGGER.debug("async_setup called with config_entry: %s", entry)
    async_setup_services(hass)
    return True


async def async_setup_entry(
    hass: HomeAssistant, config_entry: VogelsMotionMountBleConfigEntry
) -> bool:
    """Set up Vogels Motion Mount Integration from a config entry."""
    _LOGGER.debug("async_setup_entry called with config_entry: %s", config_entry)

    # Registers update listener to update config entry when options are updated.
    unsub_update_listener = config_entry.add_update_listener(async_reload_entry)

    # Initialise the coordinator that manages data updates from your api.
    coordinator = VogelsMotionMountBleCoordinator(
        hass=hass,
        config_entry=config_entry,
        unsub_update_listener=unsub_update_listener,
    )

    try:
        await coordinator.api.refresh_data()
    except APIConnectionDeviceNotFoundError as err:
        raise ConfigEntryNotReady("error_device_not_found") from err
    except APIAuthenticationError as err:
        raise ConfigEntryAuthFailed("error_invalid_athentication") from err
    except Exception as err:
        raise ConfigEntryError(
            translation_key=f"Something went wrong {err}",
            translation_placeholders={"error": err},
        ) from err

    config_entry.runtime_data = coordinator
    await hass.config_entries.async_forward_entry_setups(config_entry, PLATFORMS)
    return True


async def async_reload_entry(
    hass: HomeAssistant, config_entry: VogelsMotionMountBleConfigEntry
) -> None:
    """Reload config entry."""
    _LOGGER.debug(
        "async_reload_entry async_reload with pin %s", config_entry.data["conf_pin"]
    )
    await hass.config_entries.async_reload(config_entry.entry_id)


async def async_unload_entry(
    hass: HomeAssistant, config_entry: VogelsMotionMountBleConfigEntry
) -> bool:
    """Unload a config entry."""
    _LOGGER.debug("async_unload_entry")
    if unload_ok := await hass.config_entries.async_unload_platforms(
        config_entry, PLATFORMS
    ):
        _LOGGER.debug("async_unload_entry pop")
        coordinator: VogelsMotionMountBleCoordinator = config_entry.runtime_data
        await coordinator.unload()

    return unload_ok
