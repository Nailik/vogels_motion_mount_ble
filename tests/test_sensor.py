"""Tests for sensor entities."""

from unittest.mock import patch

from pytest_homeassistant_custom_component.common import (
    MockConfigEntry,
    snapshot_platform,
)
from syrupy.assertion import SnapshotAssertion

from custom_components.vogels_motion_mount_ble.sensor import (
    DistanceSensor,
    RotationSensor,
)
from homeassistant.const import Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers import entity_registry as er

from .conftest import setup_integration  # noqa: TID251

# -------------------------------
# region Setup
# -------------------------------


async def test_all_entities(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
    mock_config_entry: MockConfigEntry,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test all entities."""
    with patch(  # noqa: SIM117
        "custom_components.vogels_motion_mount_ble.PLATFORMS", [Platform.SENSOR]
    ):
        with patch.object(
            DistanceSensor, "_attr_entity_registry_enabled_default", True
        ):
            with patch.object(
                RotationSensor, "_attr_entity_registry_enabled_default", True
            ):
                await setup_integration(hass, mock_config_entry)

    await snapshot_platform(hass, entity_registry, snapshot, mock_config_entry.entry_id)
