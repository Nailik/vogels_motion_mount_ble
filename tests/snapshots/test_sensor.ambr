# serializer version: 1
# name: test_all_entities[sensor.mount_authentication-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': <EntityCategory.DIAGNOSTIC: 'diagnostic'>,
    'entity_id': 'sensor.mount_authentication',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:server-security',
    'original_name': 'Authentication',
    'platform': 'vogels_motion_mount_ble',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'authentication',
    'unique_id': 'authentication',
    'unit_of_measurement': None,
  })
# ---
# name: test_all_entities[sensor.mount_authentication-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Mount Authentication',
      'icon': 'mdi:server-security',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.mount_authentication',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '2',
  })
# ---
# name: test_all_entities[sensor.mount_ceb_bl_version-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': <EntityCategory.DIAGNOSTIC: 'diagnostic'>,
    'entity_id': 'sensor.mount_ceb_bl_version',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:alpha-v',
    'original_name': 'CEB BL Version',
    'platform': 'vogels_motion_mount_ble',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'ceb_bl_version',
    'unique_id': 'ceb_bl_version',
    'unit_of_measurement': None,
  })
# ---
# name: test_all_entities[sensor.mount_ceb_bl_version-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Mount CEB BL Version',
      'icon': 'mdi:alpha-v',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.mount_ceb_bl_version',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '1.0.0',
  })
# ---
# name: test_all_entities[sensor.mount_distance-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.mount_distance',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:ruler',
    'original_name': 'Distance',
    'platform': 'vogels_motion_mount_ble',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'current_distance',
    'unique_id': 'current_distance',
    'unit_of_measurement': None,
  })
# ---
# name: test_all_entities[sensor.mount_distance-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Mount Distance',
      'icon': 'mdi:ruler',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.mount_distance',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '100',
  })
# ---
# name: test_all_entities[sensor.mount_mcp_bl_version-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': <EntityCategory.DIAGNOSTIC: 'diagnostic'>,
    'entity_id': 'sensor.mount_mcp_bl_version',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:alpha-v',
    'original_name': 'MCP BL Version',
    'platform': 'vogels_motion_mount_ble',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'mcp_hw_version',
    'unique_id': 'mcp_hw_version',
    'unit_of_measurement': None,
  })
# ---
# name: test_all_entities[sensor.mount_mcp_bl_version-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Mount MCP BL Version',
      'icon': 'mdi:alpha-v',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.mount_mcp_bl_version',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'revA',
  })
# ---
# name: test_all_entities[sensor.mount_mcp_bl_version_2-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': <EntityCategory.DIAGNOSTIC: 'diagnostic'>,
    'entity_id': 'sensor.mount_mcp_bl_version_2',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:alpha-v',
    'original_name': 'MCP BL Version',
    'platform': 'vogels_motion_mount_ble',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'mcp_bl_version',
    'unique_id': 'mcp_bl_version',
    'unit_of_measurement': None,
  })
# ---
# name: test_all_entities[sensor.mount_mcp_bl_version_2-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Mount MCP BL Version',
      'icon': 'mdi:alpha-v',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.mount_mcp_bl_version_2',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '1.0.1',
  })
# ---
# name: test_all_entities[sensor.mount_mcp_bl_version_3-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': <EntityCategory.DIAGNOSTIC: 'diagnostic'>,
    'entity_id': 'sensor.mount_mcp_bl_version_3',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:alpha-v',
    'original_name': 'MCP BL Version',
    'platform': 'vogels_motion_mount_ble',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'mcp_fw_version',
    'unique_id': 'mcp_fw_version',
    'unit_of_measurement': None,
  })
# ---
# name: test_all_entities[sensor.mount_mcp_bl_version_3-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Mount MCP BL Version',
      'icon': 'mdi:alpha-v',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.mount_mcp_bl_version_3',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '2.0.0',
  })
# ---
# name: test_all_entities[sensor.mount_pin_settings-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': <EntityCategory.DIAGNOSTIC: 'diagnostic'>,
    'entity_id': 'sensor.mount_pin_settings',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:cloud-key',
    'original_name': 'Pin Settings',
    'platform': 'vogels_motion_mount_ble',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'pin_settings',
    'unique_id': 'pin_settings',
    'unit_of_measurement': None,
  })
# ---
# name: test_all_entities[sensor.mount_pin_settings-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Mount Pin Settings',
      'icon': 'mdi:cloud-key',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.mount_pin_settings',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '13',
  })
# ---
# name: test_all_entities[sensor.mount_rotation-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.mount_rotation',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:angle-obtuse',
    'original_name': 'Rotation',
    'platform': 'vogels_motion_mount_ble',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'current_rotation',
    'unique_id': 'current_rotation',
    'unit_of_measurement': None,
  })
# ---
# name: test_all_entities[sensor.mount_rotation-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Mount Rotation',
      'icon': 'mdi:angle-obtuse',
    }),
    'context': <ANY>,
    'entity_id': 'sensor.mount_rotation',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '5',
  })
# ---
